### PR TITLE
[CALCITE-5414] Use DateTimeUtils to correctly convert between java.sql types and unix timestamps

### PR DESCRIPTION
**Pull request by freastro**
Converting java.sql types to unix timestamps requires extra steps to also convert to the correct calendar. Unix timestamps should follow the proleptic Gregorian calendar as defined by ISO-8601. Java uses the standard Gregorian calendar for java.sql types and switches to the Julian calendar for dates before the Gregorian shift.

The DateTimeUtils class in Avatica correctly handles the calendar conversions. Calcite should use those methods since its own methods do not currently convert between calendars.

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/test/QuidemTest.java
Comment (asolimando) [inline testkit/src/main/java/org/apache/calcite/test/QuidemTest.java:296]: I am wondering if this impacts our CI which tests under different timezones (see [this](https://github.com/apache/calcite/blob/687dec0afcfab781f905b16e422bc03bd6b9209e/.github/workflows/main.yml#L142-L165) for instance).

Can you try locally to run with different timezones and see if those tests are affected?

You can use ```./gradlew assemble --no-build-cache cleanTest $yourTestsHere -Duser.timezone=$TZ``` for that.

File: core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java
Comment (asolimando) [inline core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java:981]: I'd rather split these tests into smaller units, basically wherever you have a comment saying "this tests xyz" I'd turn it into a separate `testNameXyz`, so that in case of test failures it's easier to understand at a glance what broke, and tests are generally easier to read since they are shorter.

**Review by asolimando — CHANGES_REQUESTED** [review]
The change LGTM, I particularly like the unit tests you have added, I have left some comments but nothing major, thanks for your contribution!

**Comment by asolimando** [comment]
I just noticed that we are relying here on another PR from Avatica which is not yet in (https://github.com/apache/calcite-avatica/pull/197), which incidentally fails when run on a different TZ in CI.

I have explicitly marked this dependency in Jira, it's in your best interest to mark such dependencies, I might have reviewed the other PR instead of this one.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (freastro) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2069]: I updated the javadoc on the method in `SqlFunctions` that I touched. Most of it is a copy & paste from the methods I added to `DateTimeUtils`, but I thought it would be useful to re-explain it here too.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (freastro) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2306]: I added some explanation about time zones. I'm not sure if it's sufficient or if you'd like additional explanation?

File: testkit/src/main/java/org/apache/calcite/test/QuidemTest.java
Comment (freastro) [reply testkit/src/main/java/org/apache/calcite/test/QuidemTest.java:296]: I made a change in DateTimeUtils (see CALCITE-2989) that ensures the date and time accessors in AbstractCursor have the same behavior for their getString() methods, and that required a change here to fix CoreQuidemTest. The default time zone is the local time zone, but CatchallSchema is creating `java.sql.Date`, `java.sql.Time`, and `java.sql.Timestamp` values using the UTC time zone. I verified the tests are passing in multiple timezones: UTC, America/Los_Angeles, America/New_York, Pacific/Chatham, Pacific/Samoa, and Pacific/Tongatapu.

**Comment by freastro** [comment]
@asolimando, thanks for reviewing! I've made the changes you've requested. The Avatica PR this depends on is still in review. But let me know if you have any other suggestions.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (asolimando) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2306]: Thanks @freastro, I think it's great now, good job!

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (asolimando) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2069]: Yes, it's a bit unfortunate to have to copy&paste but it's good to have the proper context (or pointers to where to get it).

**Comment by asolimando** [comment]
> @asolimando, thanks for reviewing! I've made the changes you've requested. The Avatica PR this depends on is still in review. But let me know if you have any other suggestions.

You are welcome, I have reviewed also the avatica PR, of course we will need to wait for that to land and for upgrading in Calcite, but at the moment this PR LGTM, will check again when all the aforementioned steps will be done and merge if all is still good (we will need to have green tests, for instance).

**Comment by freastro** [comment]
@asolimando, just saw your comment about not squashing commits. I'll stop doing that unless requested. If it helps, it looks like GitHub can compare between commits even if they were force-pushed.

To summarize my recent force-push changes:
* I made changes to the javadoc comments that Julian had suggested from the related CALCITE-2989.
* Updated `SqlFunctionsTest.testToIntWithSqlDateInAnsiDateRange()` and `SqlFunctionsTest.testToLongWithSqlTimestampInAnsiDateRange()` to also test legacy behavior so that the tests no longer skip certain dates.

**Review by asolimando — COMMENTED** [review]
Changes LGTM, let's wait for the merge of the Avatica PR and the dependency upgrade for Avatica in Calcite for a final look, thanks @freastro for your work so far

**Comment by julianhyde** [comment]
FYI, I have just merged https://github.com/apache/calcite-avatica/pull/197.

**Pull request closed without merging**
